### PR TITLE
Mixed Copula Refactor

### PR DESCRIPTION
**Pull request by andrjohns**
I'm opening this PR to propose a refactor to how mixed-copulas are specified, but I'm also fully supportive of revising/discarding the refactor if others disagree.

Firstly, I'm proposing changing the `lpdf` name from `mixed_cop_lp` to `gaussian_copula_cholesky_lpdf`, since it feels like that's a bit more descriptive and self-explanatory.

Secondly, the `mixed_cop_lp` function currently handles both the calculation/construction of the marginals, as well as the overall likelihood. This is a bit restrictive, since it requires that the user supplies compatible 'dummy' inputs for any outcome type that isn't used in the model. I think it would be more flexible to pass the computed marginals & bounds to `mixed_cop_lp`, so that users can work with any combination of acceptable input types, and it's easier to add functions for additional margins.

To put it concretely, each outcome type would have a `marginal` function which returns the continuous outcome, and bounds for all of those outcome types for a given individual:
```stan
matrix normal_marginal(row_vector y, row_vector mu_glm, vector sigma) {
  int J = num_elements(y);
  matrix[J, 5] rtn;
  rtn[ : , 1] = Phi_approx(transpose(y - mu_glm) ./ sigma);
  rtn[ : , 2] = rep_vector(negative_infinity(), J);
  rtn[ : , 3] = rep_vector(positive_infinity(), J);
  rtn[ : , 4] = rep_vector(0, J);
  rtn[ : , 5] = rep_vector(0, J);

  return rtn;
}
```

The returned matrix then contains all information that the copula lpdf needs. To work with multiple margin types, the user would just concatenate (row-wise) the returned matrices. This allows the margins and correlations to be handled separately (much like a copula itself!):

```stan
real gaussian_copula_cholesky_lpdf(matrix marginals, vector tmvn_mu, matrix L) {
  int J = num_elements(tmvn_mu);
  vector[J] u = marginals[ : , 1];
  vector[J] lb = marginals[ : , 2];
  vector[J] ub = marginals[ : , 3];
  vector[J] lb_ind = marginals[ : , 4];
  vector[J] ub_ind = marginals[ : , 5];

  return multi_normal_cholesky_truncated_lpdf(u | tmvn_mu, L, lb, ub, lb_ind, ub_ind);
}
```

Finally, I'm also proposing to slightly re-order the input arguments, such that the `tmvn_mu` mean vector is before the cholesky factor, so that it is more consistent with the other multivariate distributions in stan:
```stan
  marginals ~ gaussian_copula_cholesky(mu_zero, L);
```

### CODE REVIEW

**Comment by andrjohns** [comment]
A possible more user-friendly alternative than requiring the user concatenate the marginal matrices would also be to have the copula likelihood take an array of matrices, which are then concatenated internally:
```stan
{normal_marginal(Yn[n], mu_n[n], sigma),
 bernoulli_marginal(Yb[n], mu_b[n], uraw_b[n]),
 poisson_marginal(Yp[n], mu_p[n], uraw_p[n])}  ~ gaussian_copula_cholesky(mu_zero, L);
```

**Comment by spinkney** [comment]
I'm in favor! Since the latest release we can use multi-dispatch. So we can add typical variants of signatures for the marginals. For eg, the normal_marginal could have a real or vector `sigma`. 

I like the array of matrices idea! In that case should we remove `mu_zero` from the gaussian copula and make it `centered_gaussian_copula_cholesky`? 

Tarheel pointed out that the bernoulli case can be vectorized by
Screen Shot 2022-03-09 at 6 53 48 AM

which gives the very simple function
```stan
matrix get_bounds_bernoulli(vector y, vector mu) {
  matrix[size(y), 2] bounds;
  vector mu_y = mu .* y;
  bounds[, 1] = y - mu_y;
  bounds[, 2] = 1 - mu + mu_y;
  return bounds;
}
```
For the other discrete margins we can create indexes for the 0 and not 0 cases. I did it for the bernoulli case before the above code but should be easy to adapt. The idea is
```stan
functions {
    int [] sum_bernoulli (int[,] Y) {
    int num_cols = dims(Y)[2];
    int sums[num_cols];
    
    for (i in 1:num_cols) 
      sums[i] = sum(Y[, i]);
      
    return sums;
  }
  
  int [,] bernoulli_index (int[,] Y) {
    int num_rows = dims(Y)[1];
    int num_cols = dims(Y)[2];
    int bernoulli_totals[num_cols] = sum_bernoulli(Y); 
    int indx_bernoulli[num_rows, num_cols] = rep_array(0, num_rows, num_cols);
  
    int counter[num_cols, 2] = rep_array(0, num_cols, 2);
    
    for (n in 1:num_rows){
      for (j in 1:num_cols) {
        if (Y[n, j] == 1){
          counter[j, 1] += 1;
          indx_bernoulli[counter[j, 1], j] = n;
        } else {
          counter[j, 2] += 1;
          indx_bernoulli[bernoulli_totals[j] + counter[j, 2], j] = n;
        }
      }
    }
    
    return indx_bernoulli;
  }
}
data {
  int<lower=0> N;
  int Jb;
   int<lower=0,upper=1> Yb[N, Jb]; 
}
parameters {
  real mu;
}
model {
  mu ~ std_normal();
}
generated quantities {
  int bernoulli_totals[Jb] = sum_bernoulli(Yb); 
  int indx_bernoulli[N, Jb] = bernoulli_index(Yb);
}
```
we can compare to R
```r
fp <- file.path("~/test_index.stan")
mod <- cmdstan_model(fp)
# Run one chain of the model
num_berns <- 10
berns <- lapply(1:num_berns, function(x) rbinom(200, 1, rbeta(1, 2, 2)))
Yb <- matrix(unlist(berns), ncol = num_berns, byrow = TRUE)

test <- mod$sample(data = list(N = 200,
                              Jb = num_berns,
                              Yb = Yb), chains = 1, parallel_chains = 4)

stan_out <- matrix(test$summary("indx_bernoulli")$mean, nrow = 200, ncol = num_berns)
r_out <- matrix(unlist(lapply(1:num_berns, function(x) c( which(Yb[, x] == 1), which(Yb[, x] == 0)))), nrow = 200, ncol = num_berns)

all.equal(stan_out, r_out)
```

**Comment by andrjohns** [comment]
> In that case should we remove mu_zero from the gaussian copula and make it centered_gaussian_copula_cholesky?

Agreed! The `mu_zero` feels like it would be a source of confusion very easily.

> Tarheel pointed out that the bernoulli case can be vectorized by

Oh wow, that's well clever. I can introduce that here.

> For the other discrete margins we can create indexes for the 0 and not 0 cases.

Sorry I got lost here, what can we do if we have the indexes for non-zero cases?

**Comment by spinkney** [comment]
Forget the index stuff for 0. That only works with vectorized cdfs. If we had vectorized cdfs then we could remove the conditional `if` and supply indexes for when the vector is 0 and not, which we would supply as `lb` and `ub`. Given that we're introducing `qf` and possibly copulas in Stan, it may make sense to have a signature like `*_vcdf` or something to indicate a vector return for the number of observations.

**Comment by dmi3kno** [comment]
I think you should check with @bgoodri regarding the naming convention. In my own functions I differentiate between vectorized and scalar functions by an additional suffix, i.e. for CDF: `mydist_v_cdf()`

**Comment by spinkney** [comment]
I remembered I had a bunch of notes on my conversation with tarheel and there's a few more things that we should do. In fact, we don't need the truncated mvn. Tarheel showed that this is equivalent. I believe your code has functionalized all the marginals. We just replace the tmvn with the `multi_normal_cholesky_copula_lpdf` (which we also spent time optimizing.

```stan
functions{ 
 real multi_normal_cholesky_copula_lpdf(matrix U, matrix L) {
      int N                  = rows(U);
      int J                  = cols(U);
     matrix[N,J] Q          = to_matrix( inv_Phi(to_vector(U)), N, J );
      matrix[J,J] Gammainv   = chol2inv(L);
      return -N * sum(log(diagonal(L))) - 0.5 * sum( add_diag(Gammainv, -1.0) .* crossprod(U) );
   }

}
data {
  int<lower=0> N;                 // number of observations
  int<lower=0> J;                 // total number of outcomes
  int<lower=0> Jn;                // number of normal outcomes
  int<lower=0> Jb;                // number of binomial outcomes
  int<lower=0> Jp;                // number of poisson outcomes
  int<lower=0> K;                 // number of covariates for concatenated design matrices (X1, .., XK)
  matrix[N, Jn]        Yn;        // normal outcomes
  int<lower=0,upper=1> Yb[N, Jb]; // bernoulli outcomes
  int<lower=0>         Yp[N, Jp]; // poisson outcomes
  matrix[N,K]  X;                 // concatenated design matrices (X1, ..., XK)
  int<lower=0> Xindx[J,2];        // Jx2 integer array giving the start and end indexes of X matrix for each outcome
  int<lower=0> Kj[J];             // J-dim integer array giving how many covariates per outcome
  vector[K] y0j_Xj;               // K-dim row vector giving y_{0j}'X_j for j = 1, ..., J; y_{0j} = prior prediction
  vector[J] lambda;               // precision parameters for conjugate prior
}
parameters {
  vector[K] beta;                             // long vector of all regression coefficients
  cholesky_factor_corr[J] L;                  // Cholesky decomposition of JxJ correlation matrix
  real<lower=0> sigmasq[Jn];                  // Jn-dim array of variances (may be 0-dim)
  matrix<lower=0,upper=1>[N,Jb] uraw_b;       // latent variables for bernoulli outcomes
  matrix<lower=0,upper=1>[N,Jp] uraw_p;       // latent variables for Poisson outcomes
}
model {
  // define helper quantities
  int j  = 1;    // counter for outcomes
  int jn = 1;    // counter for normal outcomes
  int jb = 1;    // counter for bernoulli outcomes
  int jp = 1;    // counter for poisson outcomes
  matrix[N,J] U; // matrix of copula random variables
  
  // ADD LOG CAUCHY PRIOR FOR NORMAL DISPERSION PARAMETER
  if ( Jn > 0 )
    sigmasq ~ cauchy(0.0, 20.0);
  
  // FOR NORMAL RESPONSES
  //    1. ADD LOG LIKELIHOOD TO TARGET
  //    2. ADD LOG PRIOR TO TARGET
  //    3. CONSTRUCT UNIFORM RANDOM VARIABLE: 
  //       U = PHI(Z); Z = (Y - MU) / SIGMA ~ N(0,1)
  while ( jn <= Jn ) {
    int start                   = Xindx[j,1];        // start index of design matrix
    int end                     = Xindx[j,2];        // end index of design matrix
    matrix[N,Kj[j]] X_j         = X[:,start:end];    // design matrix for jth outcome
    vector[Kj[j]] beta_j        = beta[start:end];   // subset of beta pertaining to j-th outcome
    vector[Kj[j]] y0tilde_j     = y0j_Xj[start:end]; // y_{0j}'X_j
    vector[N]     mu_j          = X_j * beta_j;      // mean of jth outcome
    vector[N]     y_j           = Yn[:,jn];          // jth response variable
    
    // normal likelihood
    y_j ~ normal(mu_j, sqrt(sigmasq[j]));
    
    // conjugate linear model prior
    target += lambda[j] * inv(sigmasq[j]) * (sum(y0tilde_j .* beta_j) - 0.5 * sum(square(mu_j)) );
    
    // Create uniform random variable
    U[:,j] = Phi_approx( inv_sqrt(sigmasq[j]) * (y_j - mu_j) );
    
    // Increment counters for normal and all outcomes
    jn += 1;
    j  += 1;
  }
  
  // FOR BERNOULLI RESPONSES
  //    1. COMPUTE BOUNDS
  //        IF Y == 0, THERE IS AN UPPER BOUND AT F(0) = 1 - MU AND LOWER BOUND AT 0
  //        IF Y == 1, THERE IS A  LOWER BOUND AT F(0) = 1 - MU AND UPPER BOUND AT 1
  //    2. SET U[i,j] = LB + (UB - LB) U_RAW ~ U(LB, UB)
  //    3. JACOBIAN ADJUSTMENTS
  //          ADD LOG(UB - LB) TO TARGET
  //    4. ADD LOG CONJUGATE PRIOR FOR REGRESSION COEFFICIENTS
  while ( jb <= Jb ) {
    int start               = Xindx[j,1];                   // start index of design matrix
    int end                 = Xindx[j,2];                   // end index of design matrix
    matrix[N,Kj[j]] X_j     = X[:,start:end];               // design matrix for jth outcome
    vector[Kj[j]] beta_j    = beta[start:end];              // subset of beta pertaining to j-th outcome
    vector[Kj[j]] y0tilde_j = y0j_Xj[start:end];            // y_{0j}'X_j
    vector[N]     mu_j      = inv_logit( X_j * beta_j );    // mean of jth outcome
    
    // construct uniform random variables and jacboian adjustments
    for ( i in 1:N ) {
      real Lbound = 0;
      real Ubound = 1;
      real UmL = 0;
      if ( Yb[i,jb] == 0 )
        Ubound = 1 - mu_j[i];
      else
        Lbound = 1 - mu_j[i];
      UmL = Ubound - Lbound;
      U[i,j] = Lbound + UmL * uraw_b[i,jb];
      target += log(UmL);
    }
    
    // conjugate logistic regression prior
    target += lambda[j] * ( sum(y0tilde_j .* beta_j) + sum(log(1 - mu_j) ) );
    
    // Increment counters for Bernoulli and all outcomes.
    jb += 1;
    j  += 1;
  }
  
  
  // FOR POISSON RESPONSES
  //    1. COMPUTE BOUNDS
  //        THERE IS ALWAYS AN UPPER BOUND (UB) AT F(Y | MU)
  //        IF Y == 0, LOWER BOUND (LB) AT 0
  //        IF Y > 0, LB AT F(Y-1 | MU)
  //    2. GENERATE U[i,j] ~ U(LB, UB)
  //          U = LB + (UB - LB) * U_RAW ~ U(LB, UB)
  //    3. JACOBIAN ADJUSTMENTS
  //          ADD LOG(UB - LB) TO TARGET
  //    4. ADD LOG CONJUGATE PRIOR FOR REGRESSION COEFFICIENTS
    while ( jp <= Jp ) {
    int start               = Xindx[j,1];                   // start index of design matrix
    int end                 = Xindx[j,2];                   // end index of design matrix
    matrix[N,Kj[j]] X_j     = X[:,start:end];               // design matrix for jth outcome
    vector[Kj[j]] beta_j    = beta[start:end];              // subset of beta pertaining to j-th outcome
    vector[Kj[j]] y0tilde_j = y0j_Xj[start:end];            // y_{0j}'X_j
    vector[N]     mu_j      = exp( X_j * beta_j );          // mean of jth outcome
    
    // construct uniform random variables and jacboian adjustments
    for ( i in 1:N ) {
      real Ubound = poisson_cdf(Yp[i,jp], mu_j[i]);
      real Lbound = 0;
      real UmL = 0;
      if ( Yp[i,jp] > 0 )
        Lbound = poisson_cdf(Yp[i,jp] - 1, mu_j[i]);
      UmL    = Ubound - Lbound;
      U[i,j] = Lbound + UmL * uraw_p[i,jp];
      target += log(UmL);
    }
    
    // conjugate poisson regression prior
    target += lambda[j] * ( sum(y0tilde_j .* beta_j) - sum(mu_j) );
    
    // Increment counter for Poisson and all outcomes
    jp += 1;
    j  += 1;
  }
  
  // Increase target probability with Gaussian copula density
  L ~ lkj_corr_cholesky(1.0);
  U ~ multi_normal_cholesky_copula(L)
}
generated quantities {
  corr_matrix[J] Gamma = multiply_lower_tri_self_transpose(L);
}

```

**Comment by andrjohns** [comment]
Oh nice. So the key is to have both the `multi_normal_cholesky_copula` and the jacobian adjustment for each RV? Then in that case the individual `_marginal` functions just have to return the random variable and the jacobian adjustment, which is much cleaner! I'll get that worked up

**Comment by andrjohns** [comment]
I'm having trouble getting equivalent results with the new likelihood unfortunately. I've changed the `_marginal` functions to take the full matrix/array of responses for all individuals and return an array of matrices, where the first matrix has the constructed uniform RV's and the second matrix has the associated jacobian adjustment. This leads to much cleaner code where the likelihood for all individuals and responses is specified as:
```stan 
  { normal_marginal(Yn, mu_n, sigma),
    bernoulli_marginal(Yb, mu_b, uraw_b),
    poisson_marginal(Yp, mu_p, uraw_p) } ~ centered_gaussian_copula_cholesky(L);
```

Unfortunately the results are not equivalent, where the estimated correlation matrix for the current (old) specification is:
```r
# A tibble: 9 × 10
  variable      mean  median     sd    mad     q5    q95   rhat ess_bulk ess_tail
  <chr>        <dbl>   <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>    <dbl>    <dbl>
1 Gamma[1,1]  1       1      0      0       1     1      NA          NA       NA 
2 Gamma[2,1]  0.331   0.334  0.0522 0.0490  0.247 0.414   1.00     1899.     610.
3 Gamma[3,1]  0.286   0.289  0.0458 0.0461  0.207 0.357   0.998    1885.     451.
4 Gamma[1,2]  0.331   0.334  0.0522 0.0490  0.247 0.414   1.00     1899.     610.
5 Gamma[2,2]  1       1      0      0       1     1      NA          NA       NA 
6 Gamma[3,2] -0.0498 -0.0450 0.0641 0.0638 -0.158 0.0491  1.01     1220.     635.
7 Gamma[1,3]  0.286   0.289  0.0458 0.0461  0.207 0.357   0.998    1885.     451.
8 Gamma[2,3] -0.0498 -0.0450 0.0641 0.0638 -0.158 0.0491  1.01     1220.     635.
9 Gamma[3,3]  1       1      0      0       1     1      NA          NA       NA 
```
While the new likelihood has:
```r
# A tibble: 9 × 10
  variable    mean median      sd     mad    q5   q95  rhat ess_bulk ess_tail
  <chr>      <dbl>  <dbl>   <dbl>   <dbl> <dbl> <dbl> <dbl>    <dbl>    <dbl>
1 Gamma[1,1] 1      1     0       0       1     1     NA         NA       NA 
2 Gamma[2,1] 0.967  0.967 0.00308 0.00306 0.962 0.972  1.00     444.     527.
3 Gamma[3,1] 0.956  0.956 0.00351 0.00348 0.950 0.961  1.00     711.     655.
4 Gamma[1,2] 0.967  0.967 0.00308 0.00306 0.962 0.972  1.00     444.     527.
5 Gamma[2,2] 1      1     0       0       1     1     NA         NA       NA 
6 Gamma[3,2] 0.944  0.944 0.00478 0.00477 0.936 0.952  1.00     709.     697.
7 Gamma[1,3] 0.956  0.956 0.00351 0.00348 0.950 0.961  1.00     711.     655.
8 Gamma[2,3] 0.944  0.944 0.00478 0.00477 0.936 0.952  1.00     709.     697.
9 Gamma[3,3] 1      1     0       0       1     1     NA         NA       NA
```

For simplicity, I've adjusted the example model (`centered_gaussian_copula_example.stan`) to perform the concatenation and jacobian aggregations in the `model` and `transformed parameters` blocks, which will hopefully make it a bit easier to debug. Any ideas about where the discrepancy might be coming from?

**Comment by andrjohns** [comment]
Also just mentioning that all of the `_marginal` functions currently use loops, rather than any vectorisation, since I wanted to be sure that any discrepancies weren't caused by accidentally mis-specifying anything. The plan is to change that once the issues are sorted

**Comment by spinkney** [comment]
@andrjohns I think I found the issue.

The code I posted didn't include `inv_Phi` of the non-normal marginals. 

For e.g., on line 119 you need,
```stan
  rtn[1][n, j] = inv_Phi(Lbound + UmL * u_raw[n, j]);
```
Updating all those I get
```r
r$> matrix(mod_out$summary("Gamma")$mean, 3,3)                                                                                                                                                        
           [,1]       [,2]       [,3]
[1,]  1.0000000 -0.8632196 -0.5634402
[2,] -0.8632196  1.0000000  0.5562914
[3,] -0.5634402  0.5562914  1.0000000

r$> Gamma                                                                                                                                                                                             
           [,1]       [,2]       [,3]
[1,]  1.0000000 -0.5537555 -0.4456616
[2,] -0.5537555  1.0000000  0.4354889
[3,] -0.4456616  0.4354889  1.0000000
```

File: functions/copula/centered_gaussian_copula.stanfunctions [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+  for (j in 1:J) {
+    for (n in 1:N) {
+      real Ubound = poisson_cdf(y[n, j] | mu_glm_exp[n, j]);
+      real Lbound = 0;
+      if (y[n, j] > 0) {
+        Lbound = poisson_cdf(y[n, j] - 1 | mu_glm_exp[n, j]);
+      }
+      real UmL = Ubound - Lbound;
+      rtn[1][n, j] = Lbound + UmL * u_raw[n, j];
```
Comment (spinkney) [inline functions/copula/centered_gaussian_copula.stanfunctions:159]: should be
```stan
rtn[1][n, j] = inv_Phi(Lbound + UmL * u_raw[n, j]);
```

File: functions/copula/centered_gaussian_copula.stanfunctions [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+  for (j in 1:J) {
+    for (n in 1:N) {
+      real Ubound = binomial_cdf(num[n, j] | den[n, j], mu_glm_logit[n, j]);
+      real Lbound = 0;
+      if (num[n, j] > 0) {
+        Lbound = binomial_cdf(num[n, j] - 1 | den[n, j], mu_glm_logit[n, j]);
+      }
+      real UmL = Ubound - Lbound;
+      rtn[1][n, j] = Lbound + UmL * u_raw[n, j];
```
Comment (spinkney) [inline functions/copula/centered_gaussian_copula.stanfunctions:119]: change to
```stan
rtn[1][n, j] = inv_Phi(Lbound + UmL * u_raw[n, j]);
```

File: functions/copula/centered_gaussian_copula.stanfunctions [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    for (n in 1:N) {
+      real Lbound = 0;
+      real Ubound = 1;
+      if (y[n,j] == 0) {
+        Ubound = 1 - mu_glm_logit[n,j];
+      } else {
+        Lbound = 1 - mu_glm_logit[n,j];
+      }
+      real UmL = Ubound - Lbound;
+      rtn[1][n,j] = Lbound + UmL * u_raw[n,j];
```
Comment (spinkney) [inline functions/copula/centered_gaussian_copula.stanfunctions:77]: to
```stan
      rtn[1][n,j] = inv_Phi(Lbound + UmL * u_raw[n,j]);
```

**Comment by andrjohns** [comment]
Ah, but in the `multi_normal_cholesky_copula_lpdf`,  `inv_Phi` is also applied to the whole matrix in the construction of Q. Is the double `inv_Phi` intentional?

```stan
 real multi_normal_cholesky_copula_lpdf(matrix U, matrix L) {
      int N                  = rows(U);
      int J                  = cols(U);
     matrix[N,J] Q          = to_matrix( inv_Phi(to_vector(U)), N, J );
      matrix[J,J] Gammainv   = chol2inv(L);
      return -N * sum(log(diagonal(L))) - 0.5 * sum( add_diag(Gammainv, -1.0) .* crossprod(U) );
   }
```

**Comment by spinkney** [comment]
Ok, so I changed it to use the old `multi_normal_copula_lpdf` which doesn't have that `inv_Phi`. So what you need to do is take out that inverse_phi in the `multi_normal_cholesky_copula_lpdf` and only apply it to the non-normal margins (maybe it's the non-cont margins?). 

In the original code, I believe only the discrete margins were given the inv_phi

**Comment by andrjohns** [comment]
Got it! I'll make the changes and do some more testing. Thanks!

**Comment by andrjohns** [comment]
Made the changes, it seems much more consistent, but still slightly off the generating values:
```r
> mod_out$summary(variables = "Gamma")
# A tibble: 9 × 10
  variable     mean median     sd    mad     q5    q95  rhat ess_bulk ess_tail
  <chr>       <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl> <dbl>    <dbl>    <dbl>
1 Gamma[1,1]  1      1     0      0       1      1     NA         NA       NA 
2 Gamma[2,1]  0.817  0.821 0.0388 0.0373  0.741  0.873  1.01     117.     275.
3 Gamma[3,1] -0.446 -0.451 0.0666 0.0676 -0.548 -0.338  1.00     856.     788.
4 Gamma[1,2]  0.817  0.821 0.0388 0.0373  0.741  0.873  1.01     117.     275.
5 Gamma[2,2]  1      1     0      0       1      1     NA         NA       NA 
6 Gamma[3,2] -0.497 -0.500 0.0581 0.0587 -0.587 -0.397  1.01     753.     638.
7 Gamma[1,3] -0.446 -0.451 0.0666 0.0676 -0.548 -0.338  1.00     856.     788.
8 Gamma[2,3] -0.497 -0.500 0.0581 0.0587 -0.587 -0.397  1.01     753.     638.
9 Gamma[3,3]  1      1     0      0       1      1     NA         NA       NA 
> Gamma
           [,1]       [,2]       [,3]
[1,]  1.0000000  0.4327292 -0.3564908
[2,]  0.4327292  1.0000000 -0.3923700
[3,] -0.3564908 -0.3923700  1.0000000
```

Is that to be expected?

**Comment by spinkney** [comment]
No, there's either a bug or this method is flawed vs the truncated normal. Ask tarheel. I'll take a look and see if I find anything. In the meantime, if you want to move on, revert to the tmvn and I'll accept all the other changes.

**Comment by spinkney** [comment]
No, there's either a bug or this method is flawed vs the truncated normal. Ask tarheel. I'll take a look and see if I find anything. In the meantime, if you want to move on, revert to the tmvn and I'll accept all the other changes. 

Can you test the previous mvn copula lpdf? The way it's written you just need to transpose U but that's it.

**Comment by spinkney** [comment]
Ok I got this to work!

change the multi normal cholesky copula lpdf to
```stan
real multi_normal_cholesky_copula_lpdf(matrix U, matrix L) {
  int N = rows(U);
  int J = cols(U);
  matrix[J,J] Gammainv = chol2inv(L);
  return -N * sum(log(diagonal(L))) - 0.5 * sum(add_diag(Gammainv, -1.0) .* crossprod(U));
}
```
and you do need `inv_Phi` on the normal marginal
```stan
      rtn[1][n,j] = inv_Phi(Phi_approx((y[n,j] - mu_glm[n,j]) / sigma[j]));
```
then
```r
r$> Gamma                                                                                                                                                                       
           [,1]       [,2]       [,3]
[1,]  1.0000000 -0.4387004 -0.2157554
[2,] -0.4387004  1.0000000 -0.2523593
[3,] -0.2157554 -0.2523593  1.0000000

r$> mod_out$summary("Gamma")                                                                                                                                                    
# A tibble: 9 × 10
  variable     mean median     sd    mad     q5    q95  rhat ess_bulk ess_tail
  <chr>       <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl> <dbl>    <dbl>    <dbl>
1 Gamma[1,1]  1      1     0      0       1      1     NA         NA       NA 
2 Gamma[2,1] -0.446 -0.449 0.0475 0.0477 -0.521 -0.367  1.01     594.     698.
3 Gamma[3,1] -0.272 -0.273 0.0418 0.0418 -0.339 -0.203  1.01    1053.     452.
4 Gamma[1,2] -0.446 -0.449 0.0475 0.0477 -0.521 -0.367  1.01     594.     698.
5 Gamma[2,2]  1      1     0      0       1      1     NA         NA       NA 
6 Gamma[3,2] -0.234 -0.235 0.0587 0.0616 -0.329 -0.136  1.00     549.     727.
7 Gamma[1,3] -0.272 -0.273 0.0418 0.0418 -0.339 -0.203  1.01    1053.     452.
8 Gamma[2,3] -0.234 -0.235 0.0587 0.0616 -0.329 -0.136  1.00     549.     727.
9 Gamma[3,3]  1      1     0      0       1      1     NA         NA       NA 

```

**Comment by andrjohns** [comment]
Alright! Changes are made and the results are consistent, I've also updated the documentation and vectorisation of the marginals (where possible). This is ready for another look

File: examples/copula/R/centered_gaussian_copula_example.R [hunk 4]
```diff
@@ -1,8 +1,10 @@
 library(cmdstanr)
 # setwd("~\helpful_stan_functions")
-fp <- file.path("./examples/copula/stan/mixed_copula_example.stan")
-mod <- cmdstan_model(fp, include_paths = c("./functions/copula",
-                                          "./functions/distribution"))
+fp <- file.path("./examples/copula/stan/centered_gaussian_copula_example.stan")
+mod <- cmdstan_model(fp, include_paths = c("./functions",
+                                           "./functions/copula",
+                                           "./functions/distribution",
+                                           "../functions"))
```
Comment (spinkney) [inline examples/copula/R/centered_gaussian_copula_example.R:7]: the "../functions" line causes an error for me. Does your machine need this?

**Review by spinkney — COMMENTED** [review]
This looks great! Thanks Andrew!

The only thing I can think which would complete this is adding the `negative_binomial` distribution and an example of a different continuous distribution like `beta` or `gamma`. But those can be done in a separate PR.

File: examples/copula/R/centered_gaussian_copula_example.R (hunk 4 above)
Comment (andrjohns) [reply examples/copula/R/centered_gaussian_copula_example.R:7]: Ah no it doesn't, I'm not sure why I had that. Will pull out

**Comment by andrjohns** [comment]
> The only thing I can think which would complete this is adding the negative_binomial distribution and an example of a different continuous distribution like beta or gamma. But those can be done in a separate PR.

Yeah I'd prefer to leave that as a separate PR, since I'd like to put together some examples/documentation for how to construct the marginal functions for new distributions. I'll open an issue for it so that I don't forget.

That's all the changes I have, can you approve the workflows to run and then I'll fix up any issues they identify

**Pull request merged**